02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,225 @@
+  private readonly _chargingStationSecurityInfoRepository: IChargingStationSecurityInfoRepository;
+
+  private readonly _signedMeterValuesConfiguration:
+    | SignedMeterValuesConfig
+    | undefined;
+
+  /**
+   * @param {IFileAccess} [fileAccess] - The `fileAccess` allows access to the configured file storage.
+   *
+   * @param {SystemConfig} config - The `config` contains the current system configuration settings.
+   *
+   * @param {Logger<ILogObj>} [logger] - The `logger` represents an instance of {@link Logger<ILogObj>}.
+   *
+   * @param {IDeviceModelRepository} [deviceModelRepository] - The `deviceModelRepository` represents a repository for accessing and manipulating variable attribute data.
+   *
+   */
+  constructor(
+    fileAccess: IFileAccess,
+    config: SystemConfig,
+    logger: Logger<ILogObj>,
+    deviceModelRepository: IDeviceModelRepository,
+  ) {
+    this._fileAccess = fileAccess;
+    this._logger = logger;
+    this._deviceModelRepository = deviceModelRepository;
+    this._chargingStationSecurityInfoRepository =
+      new sequelize.SequelizeChargingStationSecurityInfoRepository(
+        config,
+        logger,
+      );
+
+    this._signedMeterValuesConfiguration =
+      config.modules.transactions.signedMeterValuesConfiguration;
+  }
+
+  /**
+   * Checks the validity of a meter value.
+   *
+   * If a meter value is unsigned, it is valid.
+   *
+   * If a meter value is signed, it is valid if:
+   * - SignedMeterValuesConfig is configured
+   * AND
+   * - The incoming signed meter value's signing method matches the configured signing method
+   * AND
+   * - The incoming signed meter value's public key is empty but there is a public key stored for that charging station
+   * OR
+   * - The incoming signed meter value's public key isn't empty and it matches the configured public key
+   *
+   * @param stationId - The charging station the meter values belong to
+   * @param meterValues - The list of meter values
+   */
+  public async validateMeterValues(
+    stationId: string,
+    meterValues: [MeterValueType, ...MeterValueType[]],
+  ): Promise<boolean> {
+    let validMeterValues = true;
+
+    const publicKeyFrequencyVariableAttribute =
+      await this._deviceModelRepository.readAllByQuerystring({
+        stationId,
+        component_name: 'OCPPCommCtrlr',
+        variable_name: 'PublicKeyWithSignedMeterValue',
+      });
+
+    const shouldPublicKeyBeUsedForValidation =
+      publicKeyFrequencyVariableAttribute.length > 0 &&
+      publicKeyFrequencyVariableAttribute[0].value !== 'Never';
+
+    if (shouldPublicKeyBeUsedForValidation) {
+      for (const meterValue of meterValues) {
+        for (const sampledValue of meterValue.sampledValue) {
+          validMeterValues =
+            validMeterValues &&
+            (await this.isSignedSampledValueValid(stationId, sampledValue));
+        }
+      }
+    }
+
+    return validMeterValues;
+  }
+
+  private async isMeterValueSignatureValid(
```

Comment:
I'm a bit conflicted on the naming since on the one hand I think it would be nice if it was named as `validateSignedMeterValue` to match the other method above which is prefixed with `validate`.

But on the other hand, I see why you went with the `is` prefix since it is a singular object being validated and not multiple.

Reply:
yup i'm always back and forth on this, for consistency though i've updated all the methods in here to be `validate...`, so this is now `validteSignedMeterValueSignature`